src/groupeddataframe/grouping.jl

```diff
@@ -245,39 +337,110 @@ end
+    f2 = ntuple(length(f)) do i
+        nms = incols[i]
+        (nms isa AbstractArray ? Tuple(nms) : nms) => last(f[i])
+    end
+    # Use temporary names for columns, and rename after the fact where appropriate
+    tmpnms = Tuple(gennames(length(f)))::NTuple{length(f), Symbol}
+    fun = function(x)
+        tup = map(f2) do p
+            local incols = first(p)
+            res = incols isa Tuple ?
+                  last(p)(NamedTuple{incols}(map(c -> x[c], incols))) :
+                  last(p)(x[incols])
+            if res isa Union{AbstractDataFrame, NamedTuple, DataFrameRow, AbstractVector, AbstractMatrix}
+                throw(ArgumentError("a single value result is required when passing a vector or tuple " *
+                                    "of functions (got $(typeof(res)))"))
+            end
+            res
+        end
+        NamedTuple{tmpnms}(tup)
+    end
+    idx, valscat = _combine(allcols => fun, gd)
+    if f isa NamedTuple
+        nams = collect(propertynames(f))
+    else
+        nams = names(valscat)
+        for i in 1:ncol(valscat)
+            if f[i] isa Pair{<:Union{Symbol,Integer}}
+                 nams[i] = Symbol(names(gd.parent)[index(gd.parent)[first(f[i])]],
+                                  '_', funname(last(f[i])))
+            end
+        end
+    end
+    names!(valscat, nams)
     idx, valscat
 end
 
-function _combine(first::AbstractDataFrame, f::Function, gd::GroupedDataFrame)
-    m = size(first, 2)
-    idx = Vector{Int}()
-    initialcols = ntuple(i -> similar(first[i], 0), m)
-    cols = _combine!(first, initialcols, idx, 1, 1, f, gd, names(first))
-    valscat = DataFrame(collect(cols), names(first))
+function _combine(f::Any, gd::GroupedDataFrame)
+    if f isa Pair{<:Union{Symbol,Integer}}
+        incols = (gd.parent[first(f)],)
+        fun = last(f)
+    elseif f isa Pair
+        df = gd.parent[collect(first(f))]
+        incols = NamedTuple{Tuple(names(df))}(columns(df))
+        fun = last(f)
+    else
+        incols = nothing
+        fun = f
+    end
+    firstres = do_call(fun, gd, incols, 1)
+    idx, valscat = _combine(wrap(firstres), fun, gd, incols)
+    if f isa Pair{<:Union{Symbol,Integer}} &&
+       !isa(firstres, Union{AbstractDataFrame, NamedTuple, DataFrameRow, AbstractMatrix})
+        nam = Symbol(names(gd.parent)[index(gd.parent)[first(f)]],
+                     '_', funname(fun))
+        names!(valscat, [nam])
+    end
+    return idx, valscat
+end
+
+function _combine(first::Union{NamedTuple, DataFrameRow, AbstractDataFrame},
+                  f::Any, gd::GroupedDataFrame,
+                  incols::Union{Nothing, Tuple, NamedTuple})
+    if first isa AbstractDataFrame
+        n = 0
+        eltys = eltypes(first)
+    elseif first isa NamedTuple{<:Any, <:Tuple{Vararg{AbstractVector}}}
+        n = 0
+        eltys = map(eltype, first)
+    elseif first isa DataFrameRow
+        n = length(gd)
+        eltys = eltypes(parent(first))
+    else
```

Comment:
maybe move `NamedTuple` condition as the last condition to make it clear when `else` is posible (I understand it is only a named tuple of non-vectors).

Here also a question is what if a function returns a NamedTuple mixing vectors and scalars (which is allowed in DataFrame constructors).

Reply:
It's already the last condition. I've added a comment to make this clearer.

Currently if you mix vectors and scalars the array is treated as a scalar. I've turned it into an error so that we can decide later to recycle as in constructors if we want.

Reply:
OK